packages/go_router/lib/src/route_data.dart

```diff
@@ -129,24 +148,126 @@ abstract class GoRouteData {
+      const NoOpPage();
+
+  /// [pageBuilder] is used to build the page
+  Widget builder(
+    BuildContext context,
+    GoRouterState state,
+    Widget navigator,
+  ) =>
+      throw UnimplementedError(
+        'One of `builder` or `pageBuilder` must be implemented.',
+      );
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static ShellRoute $route<T extends ShellRouteData>({
+    required String path,
+    required T Function(GoRouterState) factory,
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) {
+    T factoryImpl(GoRouterState state) {
+      final Object? extra = state.extra;
+
+      // If the "extra" value is of type `T` then we know it's the source
+      // instance of `GoRouteData`, so it doesn't need to be recreated.
+      if (extra is T) {
+        return extra;
+      }
+
+      return (_stateObjectExpando[state] ??= factory(state)) as T;
+    }
+
+    Widget builder(
+      BuildContext context,
+      GoRouterState state,
+      Widget navigator,
+    ) =>
+        factoryImpl(state).builder(
+          context,
+          state,
+          navigator,
+        );
+
+    Page<void> pageBuilder(
+      BuildContext context,
+      GoRouterState state,
+      Widget navigator,
+    ) =>
+        factoryImpl(state).pageBuilder(
+          context,
+          state,
+          navigator,
+        );
+
+    return ShellRoute(
+      builder: builder,
+      pageBuilder: pageBuilder,
+      routes: routes,
+      navigatorKey: navigatorKey,
+    );
+  }
+
+  /// Used to cache [ShellRouteData] that corresponds to a given [GoRouterState]
+  /// to minimize the number of times it has to be deserialized.
+  static final Expando<ShellRouteData> _stateObjectExpando =
+      Expando<ShellRouteData>(
+    'GoRouteState to ShellRouteData expando',
+  );
+}
+
+/// {@template typed_route}
+/// A superclass for each typed route descendant
+/// {@endtemplate}
 @Target(<TargetKind>{TargetKind.library, TargetKind.classType})
-class TypedGoRoute<T extends GoRouteData> {
-  /// Instantiates a new instance of [TypedGoRoute].
-  const TypedGoRoute({
-    required this.path,
-    this.routes = const <TypedGoRoute<GoRouteData>>[],
+class TypedRoute<T extends RouteData> {
+  /// {@macro typed_route}
+  const TypedRoute._({
+    this.routes = const <TypedRoute<RouteData>>[],
+    this.path,
   });
 
-  /// The path that corresponds to this rout.
-  ///
-  /// See [GoRoute.path].
-  final String path;
+  /// Instantiate a [TypedRoute] with a [path] and [routes].
+  factory TypedRoute.go({
+    required String path,
+    List<TypedRoute<RouteData>> routes = const <TypedRoute<RouteData>>[],
+  }) =>
+      TypedRoute<T>._(routes: routes, path: path);
+
+  /// Instantiate a [TypedRoute] with [routes].
+  factory TypedRoute.shell({
+    List<TypedRoute<RouteData>> routes = const <TypedRoute<RouteData>>[],
+  }) =>
+      TypedRoute<T>._(routes: routes);
```

Comment:
I have two ideas there:

both:
create abstract class TypedRoute

first:
like implemented now

second:
create separate annotations for GoRoute and ShellRoute, like
```dart
class TypedShellRoute<T extends RouteData> extends TypedRoute<T> {
  /// {@macro typed_route}
  const TypedShellRoute({
    this.routes = const <TypedRoute<RouteData>>[],
  });
}

class TypedGoRoute<T extends RouteData> extends TypedRoute<T> {
  /// {@macro typed_route}
  const TypedGoRoute({
    required this.path,
    this.routes = const <TypedRoute<RouteData>>[],
  });
}
```

Reply:
@chunhtai Hi! Created a separate PR